node/node.go

```diff
@@ -274,24 +277,49 @@ func (n *Node) run(ctx context.Context) (err error) {
 			select {
 			case <-agentDone:
 				return
-			case node := <-n.notifyNodeChange:
-				// If the server is sending us a ForceRenewal State, renew
-				if node.Certificate.Status.State == api.IssuanceStateRotate {
+			case nodeChanges := <-n.notifyNodeChange:
+				if nodeChanges.Node != nil {
+					// If the server is sending us a ForceRenewal State, renew
+					if nodeChanges.Node.Certificate.Status.State == api.IssuanceStateRotate {
+						renewCert()
+						continue
+					}
+					n.Lock()
+					// If we got a role change, renew
+					role := ca.WorkerRole
+					if nodeChanges.Node.Role == api.NodeRoleManager {
+						role = ca.ManagerRole
+					}
+					if n.role == role {
+						n.Unlock()
+						continue
+					}
+					n.Unlock()
 					renewCert()
-					continue
 				}
-				n.Lock()
-				// If we got a role change, renew
-				role := ca.WorkerRole
-				if node.Role == api.NodeRoleManager {
-					role = ca.ManagerRole
-				}
-				if n.role == role {
+
+				if nodeChanges.RootCert != nil {
+					n.Lock()
+					role := n.role
 					n.Unlock()
-					continue
+
+					if role == ca.ManagerRole || bytes.Equal(nodeChanges.RootCert, securityConfig.RootCA().Certs) {
```

Comment:
Suppose the manager stops running because it is demoted, but the role change is not seen yet (this is the `<-m.RemovedFromRaft()` case in `runManager`). Would we miss a root certificate that comes just after this?